### ISSUE TITLE
MultiTenant paging

### ISSUE THREAD

**Issue by hadfield** [weaviate/weaviate#4234]
### Describe your feature request

When retrieving the tenant list from: "/v1/schema/" + collection + "/tenants" include some optional parameters to page over the list (potentially a limit and a last returned tenant or an offset).  This may require some natural ordering of the underlying shards.

The use-case is some operation that needs to iterate over all the tenants (when the list is very big and changing) and wants to handle a chunk at a time.

Also, this may overlap with some of the MT observability functionality like a tenant count (including hot/cold counts).

### Code of Conduct

- [X] I have read and agree to the Weaviate's [Contributor Guide](https://weaviate.io/developers/contributor-guide) and [Code of Conduct](https://weaviate.io/service/code-of-conduct)

**Comment by etiennedi** [comment]
Love this, I had the same thought myself a while back. I think this should be fairly easy to do with a cursor (e.g. `?after=tenant_1234`) because they're already stored by key on disk, so seeking one tenant and then loading 50 from there should be fairly easy to do.

**Comment by etiennedi** [comment]
A simple design proposal in case someone wants to pick this up.

- backward compatible: If no parameters are set, the current behavior is not changed (it returns the entire list)
- a `limit` URL parameter is added that limits the number of tenants returned
- an `after` URL parameter is added that returns the next `n` tenants after the one specified. 
    - if the tenant specified in `after` does not exist, it seeks the next match. 

Example usage:
```
GET /v1/schema/{collection_name}/tenants?limit=100&after=tenant_123
```

would return the next 100 tenants that follow after tenant `tenant_123`

**Comment by hadfield** [comment]
and if no "after" is specified, and limit is specified, then it starts with the "first" tenant via some consistent ordering.

**Comment by DataDavD** [comment]
👋 FYI (attempting) to work on this.

cheers

**Comment by DataDavD** [comment]
👋 small update: finally got around to reviewing the code more and looking over https://github.com/weaviate/weaviate/pull/1706 for inspiration. Will start actually writing code later this week/weekend. Will keep you posted with questions and/or updates.

cheers

**Comment by DataDavD** [comment]
@etiennedi the design  proposal you first proposed may need to new restricted to just a limit. Given the [comments @dirkkul made in my PR](https://github.com/weaviate/weaviate/pull/4673) having an after var may not work, currently. Let me know what y'all think. I'm happy to tweak my PR to accommodate a limit var which seems doable still.

cheers
dd

**Comment by hadfield** [comment]
The current behavior would be to retrieve and return all tenants, which as
the comment noted, could be in the millions.  So, space-wise, keeping some
form of priority queue of the after/limit results can't be worse than that.

================
Marc C. Hadfield
***@***.***

**Comment by DataDavD** [comment]
@hadfield good call out. I believe that should be relatively easy to add using the container/heap std lib package to implement a priority queue. Are [tenant names/IDs](https://github.com/weaviate/weaviate/blob/6678b8e3602ebf5ca68e9bf4d24fa89edf410f72/usecases/sharding/state.go#L62) ordinal by definition?

**Comment by DataDavD** [comment]
@hadfield good call out. However, tenant names are not inherently ordinal given you can give a tenant any name as long as it adheres to 
> A tenant name can only contain alphanumeric characters (a-z, A-Z, 0-9), underscore (_), and hyphen (-), with a length of 4 to 64 characters.
https://weaviate.io/developers/weaviate/manage-data/multi-tenancy#add-tenants

Given your statement, I think adding a `limit` param would still be a good idea and should be a simple refactor to my current PR's code changes.

**Comment by DataDavD** [comment]
@hadfield good call out. However, it seems ordering is expected here after the `after` tenant based on previous discussions but tenant names are not inherently ordinal. I don't believe there is a tenant ID or some sort of way to order tenants consistently; a tenant can be given any name as long as it adheres to 
> A tenant name can only contain alphanumeric characters (a-z, A-Z, 0-9), underscore (_), and hyphen (-), with a length of 4 to 64 characters.
https://weaviate.io/developers/weaviate/manage-data/multi-tenancy#add-tenants

Given your statement, I think adding a `limit` param would still be a good idea and should be a simple refactor to my current PR's code changes.

**Comment by hadfield** [comment]
It's pretty much pointless to add "limit" without an "after" to page through the huge list of results (or some other form of pagination).

**Comment by hadfield** [comment]
Regarding:

> A tenant name can only contain alphanumeric characters (a-z, A-Z, 0-9), underscore (_), and hyphen (-), with a length of 4 to 64 characters.

The current behavior is to return all of them, even millions.

The worst-case new behavior is just to filter down the (huge) list to a window by imposing some natural sort.

A better solution (size-wise) would be to have some form of priority queue which would only need to keep a roughly page-size queue in memory while iterating over the (huge) list and converting each tenant id into sortable form, with the queue dropping those not in the desired window.

**Comment by DataDavD** [comment]
Yea that's hood be relatively trivial to do I believe. Curious what weaviate team says about it.

**Comment by hadfield** [comment]
If the underlying data was already in a data structure that supported an ordered view (such as a priority queue), then it becomes that much more trivial.

**Comment by dirkkul** [comment]
Etienne pointed towards the (I think) right solution at the start:

> because they're already stored by key on disk, so seeking one tenant and then loading 50 from there should be fairly easy to do.

I would look up how we implement the `after` parameter for objects and then see if something similar is possible for tenants

**Comment by DataDavD** [comment]
Maybe I misunderstood what Etienne said. Let me take another look and circle back.

Sorry, this is the first time I'm contributing to weaviate and have no experience with its code prior so if I missed/misunderstood something I apologize :/

**Comment by DataDavD** [comment]
👋 I went through the `objects` `after` implementation, which makes sense (I believe; need to re-read the code again for specifics if I need). I think I'm still hung up on location of the key/id of tenant on disk. 

I had written a bunch of thoughts but scrapped them because I'm not sure they make sense now 😅 . TLDR I though maybe I can use but that appears to be [randomly generated](https://github.com/weaviate/weaviate/blob/6678b8e3602ebf5ca68e9bf4d24fa89edf410f72/usecases/sharding/state.go#L463) 

>Each shard holds data for a single tenant.

I'm still confused by this line in the doc and how that ties to the code. I'm clearly missing something as I would expect to see some sort of tenant_id like I do object_id, but outside of a test that creates a tenant_id as UUID and used as `tenant_name` in test I'm not finding much.

> Each tenancy is like a namespace, so different tenants could, in theory, have objects with the same IDs. To avoid naming problems, object IDs in multi-tenant clusters combine the tenant ID and the object ID to create an ID that is unique across tenants. 

I'll take another look later this week; just letting y'all know since I won't be able to work on this much next few days. Sorry

**Comment by dirkkul** [comment]
I think (not 100% sure because I don't work in that area too much) that each shard (and therefore tenant) corresponds to a path on disk - paths have a natural order and could be used to implement the after parameter without additional allocations. In that path there are multiple files containing the objects in binary format etc.

**Comment by DataDavD** [comment]
👍 ok that makes more sense. I'll keep y'a updated. 

Thank you for the guidance, Dirk. I appreciate it.

**Comment by dirkkul** [comment]
you're welcome! Thanks for tackling this issue

**Comment by DataDavD** [comment]
hey sorry for the delayed response.

I went through the code quite a bit and debugged through the sharding/index/state code and I believe my understanding is right. Tenant/shard names are still limited to just the tenant name specified and the ID of a shard is just the collection/class name combined with the tenant/shard name. For example, pulling the path (note the path combines index path and shard name along with the confined root dir) for a tenant/shard I setup on my local dev is: `data/newmultitenancycollection/tenantDansby` and the shard name is: `newmultitenancycollection_tenantDansby`

We should be able to still order by that; something like prioritizing on alphanumeric characters given higher priority (0) than non-alphanumeric characters (1) and the alphanumeric characters would be ordered based on their ASCII values.  The harder part for me is getting shard path/name in the GetTenants call flow that currently exists. I believe I should be able to though with a little refactor.

Obviously updating the documentation and being explicitly clear would be important.

Weaviate let me know what you think. I went through a good amount of the code, but I could have missed something, so please let me know if there is anything else I'm missing.

**Comment by nathanwilk7** [comment]
@DataDavD Thanks for all your investigation and work to explore how this could be done, I'll spend some time looking into this as well and get back to you! Will share some thoughts when I have them (could be a few days or at most a week or two)

**Comment by hadfield** [comment]
Based on the code changes here:
https://github.com/weaviate/weaviate/pull/4741
it looks like the underlying interface for tenant information was updated to facilitate this.

**Comment by dirkkul** [comment]
>  it looks like the underlying interface for tenant information was updated to facilitate this.

This is using the same code as REST, we only added a GRPC endpoint and already prepared the paramters to contain after+limit. This does not include any code to implement pagination

**Comment by DataDavD** [comment]
Yea, what Dirk said. I actually went through that PR and noticed the same.

Thanks @nathanwilk7.  Let me know what you think/find. No worries on rushing, as I may be slow≠ to respond next couple weeks, but I'll be keeping an eye and look through the code for more inspo and/or things I may have missed. 

Thanks all!

cheers
davidd

**Comment by nathanwilk7** [comment]
@DataDavD Thanks for your patience, I've been digging into this a bit as well and wanted to share my findings.

## TL;DR

I think we should implement tenant pagination by iterating through all tenants each GET request and using a priority queue (min heap) to construct the page of tenants to return (eg what @hadfield suggested). Because of the changes in `v.1.25` of Weaviate, we no longer store tenants in the `schema.db` BoltDB file. In the future, if needed we can explore other data structures or disk-based approaches.

Pros
- No code changes outside of GetTenants.
- Extra memory usage is bounded by the limit parameter to Get Tenants.
- Can easily support adding filters on tenants (eg ActivityStatus).

Cons
- When paging through tenants, the results may change between each call.
- If the number of tenants is large, calls to this endpoint could become slow and hold a long read lock on the [cluster/store.metaClass](https://github.com/weaviate/weaviate/blob/0dd87e08e15083f8ec2b2506591237e279e6903e/cluster/store/meta_class.go#L27), which may cause problems with applying class updates.
Will not work when the number of total tenants exceeds what can be handled in memory (but in the current implementation, the whole Weaviate system will not function well in this case).

Note, we need to make sure we have metrics (or add metrics) on the lock usage and tenant pagination request time to be able to diagnose issues caused by using the lock during tenant pagination.

If needed in the future, some alternatives to explore are:

- Use a B-Tree to keep track of tenants (has nice performance characteristics for tenant operations, but may have some other system-wide performance impacts and requires large code changes).
- Store and read the tenants from external storage in a performant way (vs the current way which only uses 
- Add the ability to store a temporary “ResultSet” on disk, which could then be paged through as needed by the user (if we also need this for objects/vectors/etc then this could be compelling)
- Use a [persistent map](https://pkg.go.dev/golang.org/x/tools/internal/persistent) (or similar data structure) with a cursor so that the caller sees the same results for each call and we get “copy-on-write” performance characteristics.

## Benchmarked Solutions

In the Big-O memory/runtime notations below, n = “Number of tenants” and m = “Limit number”

### Get All (Baseline) “Don’t Add Pagination”

Code: https://github.com/weaviate/weaviate/blob/ad04c15a628f7b960fb43f5db10b8d90452ba2b6/cluster/store/schema.go#L313-L321

The biggest problem with this is that we’ll hit the internal gRPC limit when trying to return ~10M-50M tenants. In addition, adding pagination is a much better user experience and users who need all the tenants can still use the non-pagination version if they choose. In addition, returning all the tenants becomes much slower than using pagination externally because of the time de/serialize them and send them over the wire.

Expected Memory Usage: O(n)

Expected Runtime: O(n)

### Sort + Binary Search

Code: https://github.com/weaviate/weaviate/blob/ad04c15a628f7b960fb43f5db10b8d90452ba2b6/cluster/store/schema.go#L323-L350

Copies the tenant names into a slice, sorts that slice, and then uses binary search to find the tenants to return. This is slow and uses too much memory.

Expected Memory Usage: O(n) [assuming slices.Sort uses O(n) memory]

Expected Runtime: O(n + log(n)) [assuming slices.Sort has O(n + log(n)) runtime]

### Priority Queue

Code: https://github.com/weaviate/weaviate/blob/ad04c15a628f7b960fb43f5db10b8d90452ba2b6/cluster/store/schema.go#L406-L533

Iterates through all the tenants, keeping track of those to be returned in a priority queue. Very memory efficient. Not particularly fast but should be able to handle iterating through ~10M tenants in <1 second. This is probably the best approach at this time since we don’t have to sacrifice system-wide performance to support tenant pagination.

Expected Memory Usage: O(m)

Expected Runtime: O(n * log(m)) [linear scan through tenants while maintaining a priority queue of size m]

### B-Tree

Code: https://github.com/weaviate/weaviate/compare/poc/tenantpagination-btree?expand=1

This has a nice tradeoff where it uses minimal memory during the GET request and is extremely fast for range queries. However, switching to this requires quite a few code changes and is much higher-risk to cause system-wide bugs and performance issues, especially because tenant lookups which are currently low-overhead would be replaced by higher overhead calls. More investigation would be needed to investigate other potential impacts (eg memory usage, etc). The B-Tree approach could be worth doing if there are other reasons we’d like tenants backed by a B-Tree or if we need to reduce the amount of time spent in the RLockGuard.

Expected Memory Usage: O(m) [per call, not exactly sure what the per-call/general memory footprint differences would be between the current map and a b-tree]

Expected Runtime: O(log(n))

## Benchmarking

Initial benchmarks have been run with:

- 1048576 HOT tenants w/ UUID names
- Limit 1000 tenants
- After “random UUID”
- Querying a non-leader node as the coordinator
- Median of 100 GET requests used for numbers
- Note that “Get All (Baseline)” does not use limit or after since it just returns all tenants.
- Benchmarking was just run on a somewhat beefy laptop.

### Numbers

Median Memory Allocated During GET (bytes).svg

![Median Memory Allocated During GET (bytes)](https://github.com/weaviate/weaviate/assets/10050520/130a6ea5-74c9-4f67-b6bd-732fe586ef6a)

This is the amount of extra memory allocated during the code which actually finds and constructs the result tenants. “Get All” uses the most memory because it creates a slice of Tenant structs for all tenants. “Sort + Binary Search” uses memory to allocate a slice of all tenant names. “Priority Queue” uses only the memory needed to maintain the priority queue of size limit and the tenants to return. I suspect B-Tree’s memory usage is primarily internal to the btree library, though haven’t investigated more yet.

![Median Internal Time for GetTenants (ns)](https://github.com/weaviate/weaviate/assets/10050520/c0030eaf-b55c-4a3e-8401-3ed2b04baa92)

This measures how long the code which actually finds and constructs the result tenants takes. “Get All” is relatively fast since it just iterates through the tenants. “Sort + Binary Search” is slow because it sorts all tenant keys. “Priority Queue” takes the same time as iterating through all the clients plus the overhead of maintaining the priority queue. “B-Tree” is designed for iterating through things in sorted order, so it’s perfect for this and very fast.

![Median External Time for GET (ns)](https://github.com/weaviate/weaviate/assets/10050520/d346e0af-d2ae-4fc6-983d-47596f2e05d7)

This measures how long the GET call takes from the external client perspective. The thing I’m trying to show here is that the external time is dominated by the internal time if pagination is used. If no pagination is used (the “Get All” case), then the time is dominated by de/serialization and sending the data over the wire (presumably this would get worse if not run locally).

## Other Disqualified Solutions

### Add A Separate Sorted Data Structure in Addition to the Tenants Map

This would effectively double the memory footprint of tenants in the current implementation and reducing memory usage is more important than the performance gains we’d see.

## Next Steps

@DataDavD Please let me know if you'd like to implement the priority-queue based pagination or explore some other approach. If so, I'll gladly review your changes and work with you to get them merged in! If needed, I'll take a shot at building off the changes in the PR so far but would prefer to get this done together if the timing works out

**Comment by DataDavD** [comment]
hey @nathanwilk7 thank you for the great, detailed writeup 💯. 

Given what you stated and previous discussions, I think it'd be best to start with a priority queue. I'd love to implement it; I had a similar implementation POC to what you provided in your POC as well, so I don't anticipate it being overly involved in implementing. 

I guess my only callout/question is just making sure the understanding that the tenant names as a comparison is fine given its not a UUID (from my review of the code, as noted above in my previous comments) although its a string so still comparable, just may be confusing to some users starting out, but I expect that if someone is using the pagination they are fully aware of their class and shard naming conventions.

**Comment by nathanwilk7** [comment]
> I guess my only callout/question is just making sure the understanding that the tenant names as a comparison is fine given its not a UUID (from my review of the code, as noted above in my previous comments) although its a string so still comparable, just may be confusing to some users starting out, but I expect that if someone is using the pagination they are fully aware of their class and shard naming conventions.

Good callout. Let's make sure that there are good docs explaining how to do tenant pagination in the OpenAPI spec, [weaviate-io](https://github.com/weaviate/weaviate-io), and eventually in the client libraries.

Keep me posted on how things go for you and if there's anything I can do to help!

**Comment by DataDavD** [comment]
hey just a heads up. I believe using logic you suggested should be using a max heap instead  since every time we pop we would pop off the minimum value in our heap so that if we had `after=T1` and `limit=2` and our slice of tenants being `[T1, T2, T3, T4]` we would return T3 and T4 and not the expected [T2, T3] if we used a min heap for the priority queue since we would end up having to pop from the priority queue once we went over the limit after hitting the all tenants after `T1`.  

I haven't really thought it out beyond that, but I believe with current iteration and max heap priority queue instead would return the result in opposite order (i.e., T3, T2) so if we wanted to return the expected order in the returned slice then it would need to be reversed before returning.

I'll push my code before I sign off for the night so it makes more sense.

**Comment by nathanwilk7** [comment]
@DataDavD Thanks for the update and for noticing it should be a max heap! Your plan sounds good, let me know if there's anything I can do to help

**Comment by DataDavD** [comment]
Thanks @nathanwilk7. I think I'm on the right path. I'll submit the PR probably sometime relatively soon and we can discuss further there.

**Comment by DataDavD** [comment]
Hey just a small update. I haven't had time to refactor and create a formal PR. I believe I'll more time later next week so I'll at least create a draft PR we can discuss further one. 

Will keep you posted. Thanks for the patience

**Comment by DataDavD** [comment]
👋 no rush, just wondering if I can get some guidance on how to handle the API design questions I had for this PR https://github.com/weaviate/weaviate/pull/5027

**Comment by DataDavD** [comment]
hey I haven't heard anything for awhile, so I'm just bumping this to see if there is any guidance. The last place I left off in the corresponding PR there were API design questions there were best suited for Weaviate to answer or at least I thought I should sync with them before making such design decisions myself.

Also, I was wondering, given the lack of movement/comms on this issue is this something that is getting "archived" or taking a backseat while existing work is done which I'm guessing may better handle this in the future?

🥂
